File: dgx/server/routers/ollama.py
import httpx
from fastapi import APIRouter, Depends, Request
from fastapi.responses import JSONResponse, StreamingResponse

from ..auth import require_auth
from ..config import get_settings

router = APIRouter()

_UNAVAILABLE = {
    "detail": "Ollama not configured. Set OLLAMA_BASE_URL in .env"
}


def _unavailable():
    return JSONResponse(_UNAVAILABLE, status_code=503)


@router.get("/ollama/models")
async def list_models(_: None = Depends(require_auth)):
    settings = get_settings()
    if not settings.OLLAMA_BASE_URL:
        return _unavailable()
    async with httpx.AsyncClient(timeout=30) as client:
        r = await client.get(f"{settings.OLLAMA_BASE_URL}/api/tags")
        return r.json()


@router.post("/ollama/pull")
async def pull_model(request: Request, _: None = Depends(require_auth)):
    settings = get_settings()
    if not settings.OLLAMA_BASE_URL:
        return _unavailable()
    body = await request.body()

    async def _stream():
        async with httpx.AsyncClient(timeout=None) as client:
            async with client.stream(
                "POST", f"{settings.OLLAMA_BASE_URL}/api/pull", content=body
            ) as r:
                async for chunk in r.aiter_bytes():
                    yield chunk

    return StreamingResponse(_stream(), media_type="application/x-ndjson")


@router.post("/ollama/generate")
async def generate(request: Request, _: None = Depends(require_auth)):
    settings = get_settings()
    if not settings.OLLAMA_BASE_URL:
        return _unavailable()
    body = await request.body()

    async def _stream():
        async with httpx.AsyncClient(timeout=None) as client:
            async with client.stream(
                "POST", f"{settings.OLLAMA_BASE_URL}/api/generate", content=body
            ) as r:
                async for chunk in r.aiter_bytes():
                    yield chunk

    return StreamingResponse(_stream(), media_type="application/x-ndjson")


@router.post("/ollama/chat")
async def chat(request: Request, _: None = Depends(require_auth)):
    settings = get_settings()
    if not settings.OLLAMA_BASE_URL:
        return _unavailable()
    body = await request.body()

    async def _stream():
        async with httpx.AsyncClient(timeout=None) as client:
            async with client.stream(
                "POST", f"{settings.OLLAMA_BASE_URL}/api/chat", content=body
            ) as r:
                async for chunk in r.aiter_bytes():
                    yield chunk

    return StreamingResponse(_stream(), media_type="application/x-ndjson")

File: dgx/server-node/src/routers/ollama.js
'use strict';

const express = require('express');
const axios = require('axios');
const { requireAuth } = require('../auth');
const config = require('../config');

const router = express.Router();

function ollamaUnavailable(res) {
  return res.status(503).json({
    detail: 'Ollama not configured. Set OLLAMA_BASE_URL in .env',
  });
}

router.get('/ollama/models', requireAuth, async (req, res) => {
  if (!config.OLLAMA_BASE_URL) return ollamaUnavailable(res);
  try {
    const r = await axios.get(`${config.OLLAMA_BASE_URL}/api/tags`, { timeout: 30_000 });
    res.json(r.data);
  } catch (err) {
    res.status(502).json({ detail: `Ollama error: ${err.message}` });
  }
});

router.post('/ollama/pull', requireAuth, express.json(), async (req, res) => {
  if (!config.OLLAMA_BASE_URL) return ollamaUnavailable(res);
  try {
    const r = await axios.post(`${config.OLLAMA_BASE_URL}/api/pull`, req.body, {
      responseType: 'stream',
      timeout: 0,
    });
    res.setHeader('Content-Type', 'application/x-ndjson');
    r.data.pipe(res);
  } catch (err) {
    res.status(502).json({ detail: `Ollama error: ${err.message}` });
  }
});

router.post('/ollama/generate', requireAuth, express.json(), async (req, res) => {
  if (!config.OLLAMA_BASE_URL) return ollamaUnavailable(res);
  try {
    const r = await axios.post(`${config.OLLAMA_BASE_URL}/api/generate`, req.body, {
      responseType: 'stream',
      timeout: 0,
    });
    res.setHeader('Content-Type', 'application/x-ndjson');
    r.data.pipe(res);
  } catch (err) {
    res.status(502).json({ detail: `Ollama error: ${err.message}` });
  }
});

router.post('/ollama/chat', requireAuth, express.json(), async (req, res) => {
  if (!config.OLLAMA_BASE_URL) return ollamaUnavailable(res);
  try {
    const r = await axios.post(`${config.OLLAMA_BASE_URL}/api/chat`, req.body, {
      responseType: 'stream',
      timeout: 0,
    });
    res.setHeader('Content-Type', 'application/x-ndjson');
    r.data.pipe(res);
  } catch (err) {
    res.status(502).json({ detail: `Ollama error: ${err.message}` });
  }
});

module.exports = router;
